java/ql/test/experimental/query-tests/security/CWE-522-DecompressionBombs/Zip4jHandler.java

```diff
@@ -0,0 +1,48 @@
+import net.lingala.zip4j.model.LocalFileHeader;
+import net.lingala.zip4j.io.inputstream.ZipInputStream;
+
+import java.io.File;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.io.FileOutputStream;
+
+public class Zip4jHandler {
+    public static void zip4jZipInputStream(InputStream inputStream) throws IOException {
+        LocalFileHeader localFileHeader;
+        int readLen;
+        byte[] readBuffer = new byte[4096];
+        try (ZipInputStream zipInputStream = new ZipInputStream(inputStream)) {
+            while ((localFileHeader = zipInputStream.getNextEntry()) != null) {
+                File extractedFile = new File(localFileHeader.getFileName());
+                try (OutputStream outputStream = new FileOutputStream(extractedFile)) {
+                    while ((readLen = zipInputStream.read(readBuffer)) != -1) { // $ hasTaintFlow="zipInputStream"
+                        outputStream.write(readBuffer, 0, readLen);
+                    }
+                }
+            }
+        }
+    }
+
+    public static void zip4jZipInputStreamSafe(InputStream inputStream) throws IOException {
+        LocalFileHeader localFileHeader;
+        int readLen;
+        byte[] readBuffer = new byte[4096];
+        try (ZipInputStream zipInputStream = new ZipInputStream(inputStream)) { 
+            while ((localFileHeader = zipInputStream.getNextEntry()) != null) {
+                File extractedFile = new File(localFileHeader.getFileName());
+                try (OutputStream outputStream = new FileOutputStream(extractedFile)) {
+                    int totallRead = 0;
+                    while ((readLen = zipInputStream.read(readBuffer)) != -1) { // $ hasTaintFlow="zipInputStream"  "this test gives a FP"
```

Comment:
```suggestion
                    while ((readLen = zipInputStream.read(readBuffer)) != -1) { // $ SPURIOUS: hasTaintFlow="zipInputStream"
```
This is how to indicate results which should not be there. When it is removed, the tooling will say that you have fixed an incorrect result. The other side is `MISSING:`, which says that there should be a result there but there isn't.